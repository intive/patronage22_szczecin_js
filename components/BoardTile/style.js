import styled from 'styled-components'

export const Tile = styled.a`
  position: relative;
  display: flex;
  flex-direction: column;
  width: 16rem;
  height: 12.5rem;
  padding: 1rem 1rem 0.75rem 1rem;
  background-color: #FFFFFF;
  box-shadow: 0 2px 6px rgba(98, 98, 98, 0.21);
  border-radius: 8px;
  cursor: pointer;
  text-decoration: none;
`

export const IconWrapper = styled.div`
  position: absolute;
  right: 0.5rem;
  top: -0.75rem;
  width: 2.5rem;
  height: 2.5rem;
  border-radius: 50%;
  background-color: #E1E1E1;
  color: #515151;
  padding: 0.5rem;
`

export const Header = styled.h4`
  z-index: 1;
  margin-bottom: 0.25rem;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5rem;
  color: #515151;
  word-break: break-word;
`
export const TileFooter = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`

export const Text = styled.p`
  font-size: 0.75rem;
  line-height: 1rem;
  color: #7E7E7E;
`

export const Date = styled(Text)`
  flex-grow: 1;
`
