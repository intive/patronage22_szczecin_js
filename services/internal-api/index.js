import axios from 'axios'

const axiosInstance = axios.create({
  baseURL: `https://${process.env.NEXT_PUBLIC_VERCEL_URL}` || 'http://localhost:3000',
  responseType: 'json',
  withCredentials: true
})

export const getBoards = async (...args) => {
  return axiosInstance.get('/api/boards', { params: Object.assign({}, ...args) })
}
