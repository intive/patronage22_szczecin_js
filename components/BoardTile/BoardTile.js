import { useContext } from 'react'

import Icon from '../Icon/Icon'
import { Date, Header, IconWrapper, Text, Tile, TileFooter } from './style'
import BoardsContext from '../../store/boards-context'
import { useTranslation } from 'next-i18next'
import ContextMenu from '../ContextMenu/ContextMenu'
import ContextMenuItem from '../ContextMenuItem/ContextMenuItem'
import { deleteBoard } from '../../services/internal-api'

export default function BoardTile ({ id, name, date, cardCount, hasPassword }) {
  const { t } = useTranslation('common')
  const boardsCtx = useContext(BoardsContext)

  async function deleteBoardHandler (id) {
    const data = await deleteBoard(id)

    if (data.status === 204) boardsCtx.reload()
  }

  return (
    <Tile>
      {hasPassword ? <IconWrapper><Icon name='lock_outlined' /></IconWrapper> : null}
      <Header>{name}</Header>
      <Date>{date}</Date>
      <TileFooter>
        <Text>{t('cardCount', { count: cardCount })}</Text>
        <ContextMenu id={id}>
          {hasPassword ? <ContextMenuItem name={t('contextMenu.removePassword')} icon='lock_outlined' /> : <ContextMenuItem name={t('contextMenu.setPassword')} icon='lock_outlined' />}
          <ContextMenuItem name={t('contextMenu.deleteBoard')} onClick={() => deleteBoardHandler(id)} icon='delete_outlined' />
        </ContextMenu>
      </TileFooter>
    </Tile>
  )
}
