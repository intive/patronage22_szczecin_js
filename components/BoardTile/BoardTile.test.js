/* globals describe, it, expect, jest */

import { render, screen, waitFor } from '@testing-library/react'
import BoardTile from './BoardTile'
import userEvent from '@testing-library/user-event'
import * as InternalApi from '../../services/internal-api'

describe('BoardTile', () => {
  const mockDeleteBoard = jest.spyOn(InternalApi, 'deleteBoard')
  const mockUpdateBoard = jest.spyOn(InternalApi, 'updateBoard')

  it('renders correctly without password', async () => {
    const Tile = render(<BoardTile id='mockId' name='Board name 1' date='1 Jan 2022' cardCount={0} />)
    await waitFor(() => {
      expect(Tile.container).toMatchSnapshot()
    })
  })

  it('renders correctly with password', () => {
    const Tile = render(<BoardTile id='mockId' name='Board name 1' date='1 Jan 2022' cardCount={0} hasPassword />)
    expect(Tile.container).toMatchSnapshot()
  })

  it('BoardTile with password should have option removePassword and should not have option setPassword', () => {
    render(<BoardTile hasPassword />)
    userEvent.click(screen.getByText('more_horiz'))
    expect(screen.getByRole('button', { name: /contextMenu.removePassword/i })).toBeVisible()
    expect(screen.queryByRole('button', { name: /contextMenu.setPassword/i })).not.toBeInTheDocument()
  })

  it('BoardTile without password should have option setPassword and should not have option removePassword', () => {
    render(<BoardTile />)
    userEvent.click(screen.getByText('more_horiz'))
    expect(screen.getByRole('button', { name: /contextMenu.setPassword/i })).toBeVisible()
    expect(screen.queryByRole('button', { name: /contextMenu.removePassword/i })).not.toBeInTheDocument()
  })

  it('should call delete option when deleteBoard button is clicked', () => {
    render(<BoardTile id='mockedId' />)
    userEvent.click(screen.getByText('contextMenu.deleteBoard'))
    expect(mockDeleteBoard).toHaveBeenCalledTimes(1)
    expect(mockDeleteBoard).toHaveBeenCalledWith('mockedId')
  })

  it('should call patch option when removePassword button is clicked', () => {
    render(<BoardTile id='mockedId' hasPassword />)
    userEvent.click(screen.getByText('contextMenu.removePassword'))
    expect(mockUpdateBoard).toHaveBeenCalledTimes(1)
    expect(mockUpdateBoard).toHaveBeenCalledWith('mockedId', { password: null })
  })

  it('should open Modal after button click and close it after click on backdrop', () => {
    render(<BoardTile />)
    expect(screen.queryByTestId('backdrop')).not.toBeInTheDocument()
    userEvent.click(screen.getByText('more_horiz'))
    userEvent.click(screen.queryByRole('button', { name: /contextMenu.setPassword/i }))
    expect(screen.getByTestId('backdrop')).toBeInTheDocument()
    userEvent.click(screen.getByTestId('backdrop'))
    expect(screen.queryByTestId('backdrop')).not.toBeInTheDocument()
  })

  it('should open Modal after button click and close it after click on cancel button', () => {
    render(<BoardTile />)
    expect(screen.queryByRole('button', { name: /buttons.cancel/i })).not.toBeInTheDocument()
    userEvent.click(screen.getByText('more_horiz'))
    userEvent.click(screen.queryByRole('button', { name: /contextMenu.setPassword/i }))
    expect(screen.getByRole('button', { name: /buttons.cancel/i })).toBeInTheDocument()
    userEvent.click(screen.getByRole('button', { name: /buttons.cancel/i }))
    expect(screen.queryByRole('button', { name: /buttons.cancel/i })).not.toBeInTheDocument()
  })

  it('should open Modal after button click and close it after click on Esc keyboard button', () => {
    render(<BoardTile />)
    expect(screen.queryByRole('button', { name: /buttons.save/i })).not.toBeInTheDocument()
    userEvent.click(screen.getByText('more_horiz'))
    userEvent.click(screen.queryByRole('button', { name: /contextMenu.setPassword/i }))
    expect(screen.getByRole('button', { name: /buttons.save/i })).toBeInTheDocument()
    userEvent.keyboard('{esc}')
    expect(screen.queryByRole('button', { name: /buttons.save/i })).not.toBeInTheDocument()
  })

  it('should open Modal after button click and close it after click on save button if password length equal 8', () => {
    render(<BoardTile />)
    expect(screen.queryByRole('button', { name: /buttons.save/i })).not.toBeInTheDocument()
    userEvent.click(screen.getByText('more_horiz'))
    userEvent.click(screen.queryByRole('button', { name: /contextMenu.setPassword/i }))
    expect(screen.getByRole('button', { name: /buttons.save/i })).toBeInTheDocument()
    userEvent.type(screen.getByPlaceholderText('Enter password'), '12345678')
    expect(screen.getByPlaceholderText('Enter password')).toHaveValue('12345678')
    expect(screen.queryByRole('button', { name: /buttons.save/i })).not.toBeDisabled()
    userEvent.click(screen.getByRole('button', { name: /buttons.save/i }))
    expect(screen.queryByRole('button', { name: /buttons.save/i })).not.toBeInTheDocument()
  })

  it('should open Modal after button click and save button to be disabled if password length less than 8', () => {
    render(<BoardTile />)
    expect(screen.queryByRole('button', { name: /buttons.save/i })).not.toBeInTheDocument()
    userEvent.click(screen.getByText('more_horiz'))
    userEvent.click(screen.queryByRole('button', { name: /contextMenu.setPassword/i }))
    expect(screen.getByRole('button', { name: /buttons.save/i })).toBeInTheDocument()
    userEvent.type(screen.getByPlaceholderText('Enter password'), '1234')
    expect(screen.getByPlaceholderText('Enter password')).toHaveValue('1234')
    expect(screen.queryByRole('button', { name: /buttons.save/i })).toBeDisabled()
    expect(screen.queryByRole('button', { name: /buttons.save/i })).toBeInTheDocument()
  })
})
