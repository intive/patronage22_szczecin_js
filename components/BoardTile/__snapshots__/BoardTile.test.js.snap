// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardTile renders correctly with password 1`] = `
<div>
  <div
    class="sc-bdvvtL fbCQOU"
  >
    <div
      class="sc-gsDKAQ bkuMRF"
    >
      <span
        class="material-icons"
      >
        lock_outlined
      </span>
    </div>
    <h4
      class="sc-dkPtRN kUDuei"
    >
      Board name 1
    </h4>
    <p
      class="sc-hKwDye sc-eCImPb eRcqw gGNFTw"
    >
      1 Jan 2022
    </p>
    <p
      class="sc-hKwDye eRcqw"
    >
      cardCount
    </p>
  </div>
</div>
`;

exports[`BoardTile renders correctly without password 1`] = `
<div>
  <div
    class="sc-bdvvtL fbCQOU"
  >
    <h4
      class="sc-dkPtRN kUDuei"
    >
      Board name 1
    </h4>
    <p
      class="sc-hKwDye sc-eCImPb eRcqw gGNFTw"
    >
      1 Jan 2022
    </p>
    <p
      class="sc-hKwDye eRcqw"
    >
      cardCount
    </p>
  </div>
</div>
`;
